eos/utils/expression-maker.cc

```diff
@@ -0,0 +1,90 @@
+ * Public License version 2, as published by the Free Software Foundation.
+ *
+ * EOS is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
+ * Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <eos/observable.hh>
+#include <eos/utils/expression.hh>
+#include <eos/utils/expression-fwd.hh>
+#include <eos/utils/kinematic.hh>
+#include <eos/utils/options.hh>
+#include <eos/utils/parameters.hh>
+#include <eos/utils/qualified-name.hh>
+
+#include <iostream>
+
+namespace eos::exp
+{
+
+    // copies one expression tree to a new tree, replacing ObservableNameExpression objects with ObservableExpression objects
+    class ExpressionMaker
+    {
+        private:
+            Parameters _parameters;
+            Kinematics _kinematics;
+            Options    _options;
+
+        public:
+            ExpressionMaker(const Parameters & parameters, const Kinematics & kinematics, const Options & options) :
+                _parameters(parameters),
+                _kinematics(kinematics),
+                _options(options)
+            {
+            }
+
+            Expression visit(const BinaryExpression & e)
+            {
+                return BinaryExpression(e.op, e.lhs.accept_returning<Expression>(*this), e.rhs.accept_returning<Expression>(*this));
+            }
+
+            Expression visit(const ConstantExpression & e)
+            {
+                return e;
+            }
+
+            Expression visit(const ObservableNameExpression & e)
+            {
+                auto kinematics_values = e.kinematics_specification.values;
+                auto kinematics_aliases = e.kinematics_specification.aliases;
+
+                // Set or alias kinematic specifications
+                for (auto const & value : kinematics_values)
+                {
+                    _kinematics.declare(value.first, value.second);
+                }
+                for (auto const & alias : kinematics_aliases)
+                {
+                    _kinematics.alias(alias.first, alias.second);
+                }
+
+                // Make observable
+                auto observable = Observable::make(
+                    e.observable_name,
+                    this->_parameters,
+                    this->_kinematics,
+                    this->_options
+                    );
+                // Clear alias map
+                _kinematics.clear_aliases();
+
+                return ObservableExpression(observable, e.kinematics_specification);
+            }
+
+            Expression visit(const ObservableExpression & e)
+            {
+                return e;
+            }
+    };
+}
```

Comment:
``ExpressionCloner`` and ``ExpressionMaker`` seem to be very similar. Can we combine them into one class?

Reply:
Good point

Reply:
All well thought, this is maybe not a good idea. The `maker` works on `ObservableNameExpression` while the `cloner` focuses on already made `ObservableExpression`. What I can do is gather the visitors into one `expression-visitors.cc` to avoid having so many files.

Reply:
Understood!